enterprise_gateway/services/kernelspecs/kernelspec_cache.py

```diff
@@ -0,0 +1,282 @@
+# Copyright (c) Jupyter Development Team.
+# Distributed under the terms of the Modified BSD License.
+"""Cache handling for kernel specs."""
+
+import os
+from watchdog.observers import Observer
+from watchdog.events import FileSystemEventHandler, FileMovedEvent
+
+from jupyter_client.kernelspec import KernelSpec
+from notebook.utils import maybe_future
+from traitlets.config import SingletonConfigurable
+from traitlets.traitlets import CBool, default
+from typing import Dict, Optional, Union
+
+
+# Simplify the typing.  Cache items are essentially dictionaries of strings
+# to either strings or dictionaries.  The items themselves are indexed by
+# the kernel_name (case-insensitive).
+CacheItemType = Dict[str, Union[str, Dict]]
+
+
+class KernelSpecCache(SingletonConfigurable):
+    """The primary (singleton) instance for managing KernelSpecs.
+
+    This class contains the configured KernelSpecManager instance upon
+    which it uses to populate the cache (when enabled) or as a pass-thru
+    (when disabled).
+
+    Note that the KernelSpecManager returns different formats from methods
+    get_all_specs() and get_kernel_spec().  The format in which cache entries
+    are stored is that of the get_all_specs() results.  As a result, some
+    conversion between formats is necessary, depending on which method is called.
+    """
```

Comment:
Based on the comment above, does it make sense to actually have a CachedKernelSpecManager(KernelSpecManager)?

Reply:
Because the kernel spec manager class can be substituted, the cache needs to _contain_ the kernelspec manager, not _be_ the kernelspec manager.  For example, `nb-conda-kernels` introduces its own `KernelSpecManager` implementation.  As a result, I choose to not make it a subclass of `KernelSpecManager`.